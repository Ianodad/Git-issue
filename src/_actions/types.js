

// auth actions 
export const SIGN_OUT = "SIGN_OUT"
export const SIGN_IN_WITH_GITHUB = "SIGN_IN_WITH_GITHUB"
export const SIGN_IN_ANONYMOUSLY = "SIGN_IN_ANONYMOUSLY"
export const GET_USER_STATE = "GET_USER_STATE"

 // api actions
export const GET_SEARCH_USER = "GET_SEARCH_USER"
export const GET_ALL_OWNER_REPOS = "GET_ALL_USER_REPOS"
export const GET_ALL_OWNER_ISSUES = "GET_ALL_OWNER_ISSUES"
export const GET_ALL_OWNER_REPO_ISSUES = "GET_ALL_OWNER_REPO_ISSUES"
export const GET_ONE_OWNER_ISSUE = "GET_ONE_OWNER_ISSUE"
export const GET_REPO_ISSUES_COMMENTS = "GET_REPO_ISSUES_COMMENTS"
export const GET_REPO_ISSUE_COMMENTS = "GET_REPO_ISSUE_COMMENTS"

// user actions
export const SELECT_USER = "SELECT_USER"