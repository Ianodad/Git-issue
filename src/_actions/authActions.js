import firebase from "../config/firebase";

import {
  SIGN_IN_ANONYMOUSLY,
  SIGN_IN_WITH_GITHUB,
  SIGN_OUT,
  GET_USER_STATE,
} from "./types";

const SignInWithGithub = (history) => async (dispatch) => {
  
  console.log("github");
  const provider = new firebase.auth.GithubAuthProvider();

  // await firebase.auth().signInWithRedirect(provider);
  try {
    const { additionalUserInfo } = await firebase
      .auth()
      .signInWithPopup(provider);
    const user = additionalUserInfo.profile;

    //  console.log(user.login)
    const userGithub = {
      login: user.login,
      avatar_url: user.avatar_url,
      html_url: user.html_url,
      id: user.id,
      name: user.name,
      repos_url: user.repos_url,
      type: user.type,
    };

    if (user) {
      localStorage.setItem('USER', JSON.stringify(userGithub));
      dispatch({
        type: SIGN_IN_WITH_GITHUB,
        payload: userGithub,
      });

      history.push('/')
      // window.location = `/`;
    }

    // if (userGithub) {
    // }
  } catch (error) {
    console.log(error);

    dispatch({
      type: SIGN_IN_WITH_GITHUB,
      payload: null,
    });
  }

  //  console.log(additionalUserInfo)
  dispatch({
    type: SIGN_IN_WITH_GITHUB,
    payload: "sign up",
  });
};

const SignInAnonymous = () => async (dispatch) => {
  await firebase.auth().signInAnonymously();

  // window.location=`/`
  dispatch({
    type: SIGN_IN_ANONYMOUSLY,
    payload: "sign in",
  });
};

const SignOut = () => async (dispatch) => {
  // console.log("sign")
  try {
    await firebase.auth().signOut();
    localStorage.removeItem("USER");

    dispatch({
      type: SIGN_OUT,
      payload: {},
    });

    window.location = `/auth`;
  } catch (error) {
    console.log(error);
  }
};


const getUserState = () => async (dispatch) => {
  const userState = await JSON.parse(localStorage.getItem("USER"));
  // console.log(await JSON.parse(localStorage.getItem("USER")))
  console.log(userState)

  if (userState) {
    dispatch({
      type: GET_USER_STATE,
      payload: userState,
    });
  }
};
// export all action functions as an object
export const authActions = {
  SignInWithGithub,
  SignInAnonymous,
  getUserState,
  SignOut,
};
