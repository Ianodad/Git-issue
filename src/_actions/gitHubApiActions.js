
import gitHubApi from "../api/gitHubApi";
import {GET_SEARCH_USER, GET_ALL_OWNER_REPOS,  GET_ALL_OWNER_ISSUES, GET_ALL_OWNER_REPO_ISSUES, GET_ONE_OWNER_ISSUE, GET_REPO_ISSUES_COMMENTS ,GET_REPO_ISSUE_COMMENTS} from "./types";



// NOTE settings
// default username == octocat
// default repo == "Hello-World"
// search user 
const getSearchUser = (query) => async (dispatch) => {

  try {
    const {data} = await gitHubApi.getSearchUser(query)
    localStorage.setItem("userSearchResult", JSON.stringify(data.items))

    dispatch({
       type: GET_SEARCH_USER,
       payload: data.items,
     });

   if (data) {
    window.location=`/search/${query}`
  }
  } catch (error) {
    console.log(error) 
  }
   

}
// this function is responsible for get all user repos.. base on github api
const getAllOwnerRepos =(owner) => async (dispatch) => {
  const {data} = await gitHubApi.getAllOwnerRepos(owner)
  // console.log(data)
  dispatch({
      type: GET_ALL_OWNER_REPOS ,
      payload: data,
    });
}
// get all repo issues.
const  getAllOwnerRepoIssues = (owner, repo) => async (dispatch) => {
    const {data}= await gitHubApi.getAllOwnerRepoIssues(owner, repo)

    dispatch({
      type: GET_ALL_OWNER_REPO_ISSUES,
      payload: data,
    });
};


// get specific issue straight from redux state..
const getFromDispatch = (number) => async (dispatch, getState) => {
  // getAllOwnerRepoIssues()
  const data = await getState().gitHubApiData.allOwnerRepoIssues;

  try {
    const result = data.find(item=>item.number == number)
    dispatch({
        type:GET_ONE_OWNER_ISSUE,
        payload: result,
      });
    
  } catch (error) {
    console.log(error)  
  }
  // console.log(data);
  // console.log(result)

  
}

export const gitHubApiActions = {
    getAllOwnerRepoIssues,
    getAllOwnerRepos,
    getFromDispatch,
    getSearchUser
};