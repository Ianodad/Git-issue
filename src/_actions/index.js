export * from './gitHubApiActions'
export * from './authActions'
