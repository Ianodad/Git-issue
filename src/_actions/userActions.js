import {SELECT_USER} from "./types";
export * from './gitHubApiActions'



const selectUser = (user) => {
    
}
