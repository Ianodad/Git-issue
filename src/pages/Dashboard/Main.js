import React, { Component } from 'react';


import Layout from '../../Layouts/Layout';
import Card from '../../components/Card/Card.js';
import Table from '../../components/Table';
import { gitHubApiActions } from "../../_actions";

import { connect } from "react-redux";
var _ = require('lodash');


// import githubApi from "../../api/githubApi";
// github api action imported from actions 
const { getAllOwnerRepoIssues, getAllOwnerRepos }= gitHubApiActions;

class Main extends Component {
  constructor(props) {
    super(props);
    // Don't call this.setState() here!
    this.state = {};
   
  }

  componentDidMount = async () => {

    const owner = this.props.match.params.name
    ? this.props.match.params.name
    : JSON.parse(localStorage.getItem("selectUser")) ? JSON.parse(localStorage.getItem("selectUser")) :JSON.parse(localStorage.getItem("USER")).login; 
    this.props.getAllOwnerRepoIssues();
    this.props.getAllOwnerRepos(owner);

    // this.setState({data: await this.props.getAllOwnerRepoIssues() })
  };

  sumBy = (data) => {
    return _.sumBy(data, (repo) => {
      return repo.open_issues;
    });
  };

  countBy = (data)=>{
    return  _.pick(_.countBy(data, 'has_issues'),  [true])

  }
  render() {
    const { allOwnerRepos, history} = this.props;
    const repoCount = allOwnerRepos.length;
    const issue_count = this.sumBy(allOwnerRepos);
    const countSum = this.countBy(allOwnerRepos)
    const issuePercent = (100 * countSum.true)/allOwnerRepos.length
    console.log(allOwnerRepos)
    console.log(history)
    // console.log(allOwnerRepos)
    // console.log(allOwnerRepos[0].owner.login)
    if (!allOwnerRepos) {
      return (
        <Layout>
          <div>Nothing to see here</div>
        </Layout>
      );
    }
    return (
      <div>
        <Layout>
          <div className="container-fluid mt-4">
            <div className="username">
              <p className="mb-0 pb-0" style={{ color: "#C3C3E5" }}>
                Selected user
              </p>
              <a href={allOwnerRepos[0].owner} className="mt-0 pt-0">
                <h4
                  className="mt-0"
                  style={{
                    fontWeight: "bold",
                    fontSize: "40px",
                    color: "#443266",
                  }}
                >
                  {allOwnerRepos[0].owner.login}
                </h4>
              </a>
            </div>
            <div className="repo-info row mt-4 pt-4 d-flex wrap">
              <Card
                title={"Repos"}
                count={repoCount}
                style={{
                  width: "18em",
                  backgroundColor: "#C3C3E5",
                  margin: "0px 10px",
                }}
              />
              <Card
                title={"Open Issue Count"}
                count={issue_count}
                style={{ width: "18em", backgroundColor: "#C3C3E5" }}
              />
              <Card
                title={"Repository with Issues"}
                count={issuePercent.toFixed(2)+"%"}
                style={{ width: "18em", backgroundColor: "#C3C3E5" }}
              />
            </div>
          </div>
          <div className="issue mt-4">
            {allOwnerRepos && (
              <Table main={true} data={allOwnerRepos} style={{ width: "80%" }} />
            )}
          </div>
        </Layout>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  // console.log(state.gitHubApiData)
  return {
    allOwnerRepos: state.gitHubApiData.allOwnerRepos,
    repoIssues: state.gitHubApiData.allOwnerRepoIssues,
  };
};

export default connect(mapStateToProps, {
  getAllOwnerRepoIssues,
  getAllOwnerRepos
})(Main);
