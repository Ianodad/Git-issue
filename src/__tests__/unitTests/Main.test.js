import React from "react";
import { BrowserRouter, Switch } from 'react-router-dom';

import { render, fireEvent, cleanup } from "@testing-library/react";
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

import Main from "../../pages/Dashboard/Main";

import thunk from 'redux-thunk';
import {combineReducers} from 'redux';
import rootReducer from '../../_reducers';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { gitHubApiActions } from "../../_actions";
import {GET_SEARCH_USER, GET_ALL_OWNER_REPOS, GET_ALL_OWNER_ISSUES, GET_ALL_OWNER_REPO_ISSUES, GET_ONE_OWNER_ISSUE, GET_REPO_ISSUES_COMMENTS ,GET_REPO_ISSUE_COMMENTS } from "../../_actions/types";

const { getAllOwnerRepoIssues, getAllOwnerRepos }= gitHubApiActions;


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const getAllOwnerRepo = () => ({ type: GET_ALL_OWNER_REPOS })


describe("<Main/> testing component", () => {
  let store;
  let wrapper;    
  let state = {
    gitHubApiData:{
      repoIssues:"",
      allOwnerRepos:"",
      allOwnerRepoIssues: "",
      // oneOwnerIssue: "",
      }
    }
    beforeEach(() => {
    store = mockStore(state);

     // Dispatch the action
    // store.dispatch(getAllOwnerRepoIssues())
     store.dispatch = (getAllOwnerRepos());

    wrapper = render(
      <Provider store={store}>
        <BrowserRouter>
          <Switch>
            <Main />
          </Switch>
        </BrowserRouter>
      </Provider>
    );
  });

    describe("Check component renders", () => {
         it('should render with given state from Redux store', () => {
          expect(Wrapper).toMatchSnapshot();
        });
    });
});
