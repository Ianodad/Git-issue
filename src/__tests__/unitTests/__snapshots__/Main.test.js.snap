// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main/> testing component Check component renders should render with given state from Redux store 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container-fluid"
        id="layout"
      >
        <div
          class="col-md-12 row py-0"
          id="top-section"
          style="height: 60px;"
        >
          <div
            class="col-md-2 col-12 row mx-auto py-2"
            id="layout-logo"
          >
            <div
              class="col-2 my-auto"
            >
              <h1
                class="mx-auto"
                id="logo-main"
                style="font-size: 3vh;"
              >
                gitIssue
              </h1>
            </div>
          </div>
          <div
            class="col-md-10 layout"
            id="layout-menu"
          >
            <nav
              class="navbar navbar-expand-md  mt-1"
              style="height: 55px;"
            >
              <div
                class="collapse navbar-collapse col"
                id="navbarCollapse"
              >
                <form
                  class="form-inline"
                  role="form"
                >
                  <div
                    class="form-group"
                  >
                    <input
                      class="form-control"
                      placeholder="Search"
                      required=""
                      type="text"
                    />
                  </div>
                  <div
                    class="input-group-append ml-1"
                  >
                    <button
                      class="btn btn-block"
                      style="background-color: rgb(68, 50, 102); color: white;"
                      type="submit"
                    >
                      <svg
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 512 512"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                        />
                      </svg>
                    </button>
                  </div>
                </form>
              </div>
              <div
                class="navbar-nav"
              >
                <p
                  class="nav-item col"
                >
                  Username
                </p>
                <a
                  class="nav-item col"
                  href="#"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g>
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                      <path
                        d="M5 22a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1h14a1 1 0 0 1 1 1v3h-2V4H6v16h12v-2h2v3a1 1 0 0 1-1 1H5zm13-6v-3h-7v-2h7V8l5 4-5 4z"
                      />
                    </g>
                  </svg>
                </a>
              </div>
            </nav>
          </div>
        </div>
        <div
          class="col-md-12 row "
          id="bottom-section"
        >
          <div
            class="col-md-2 px-0"
            id="layout-sidemenu"
            style="height: 703px;"
          >
            <div
              class="list-group borderless mt-4 border-0 mx-0"
              id="sidebar"
            >
              <a
                aria-current="page"
                class="list-group-item list-group-item-action border-0 bg-transparent active"
                href="/"
                style="color: grey;"
              >
                <i
                  class="fa fa-home"
                />
                 Main
              </a>
              <a
                class="list-group-item list-group-item-action border-0 bg-transparent "
                href="/issues"
                style="color: rgb(140, 72, 159);"
              >
                <i
                  class="fa fa-camera"
                />
                 Issues
              </a>
              <a
                class="list-group-item list-group-item-action border-0 bg-transparent"
                href="/inspection"
                style="color: rgb(140, 72, 159);"
              >
                <i
                  class="fa fa-music"
                />
                 Inspection
              </a>
            </div>
          </div>
          <div
            class="col-md-10"
            id="layout-dashboard overflow-auto"
            style="height: 703px;"
          >
            <div>
              Nothing to see here
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container-fluid"
      id="layout"
    >
      <div
        class="col-md-12 row py-0"
        id="top-section"
        style="height: 60px;"
      >
        <div
          class="col-md-2 col-12 row mx-auto py-2"
          id="layout-logo"
        >
          <div
            class="col-2 my-auto"
          >
            <h1
              class="mx-auto"
              id="logo-main"
              style="font-size: 3vh;"
            >
              gitIssue
            </h1>
          </div>
        </div>
        <div
          class="col-md-10 layout"
          id="layout-menu"
        >
          <nav
            class="navbar navbar-expand-md  mt-1"
            style="height: 55px;"
          >
            <div
              class="collapse navbar-collapse col"
              id="navbarCollapse"
            >
              <form
                class="form-inline"
                role="form"
              >
                <div
                  class="form-group"
                >
                  <input
                    class="form-control"
                    placeholder="Search"
                    required=""
                    type="text"
                  />
                </div>
                <div
                  class="input-group-append ml-1"
                >
                  <button
                    class="btn btn-block"
                    style="background-color: rgb(68, 50, 102); color: white;"
                    type="submit"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                      />
                    </svg>
                  </button>
                </div>
              </form>
            </div>
            <div
              class="navbar-nav"
            >
              <p
                class="nav-item col"
              >
                Username
              </p>
              <a
                class="nav-item col"
                href="#"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M0 0h24v24H0z"
                      fill="none"
                    />
                    <path
                      d="M5 22a1 1 0 0 1-1-1V3a1 1 0 0 1 1-1h14a1 1 0 0 1 1 1v3h-2V4H6v16h12v-2h2v3a1 1 0 0 1-1 1H5zm13-6v-3h-7v-2h7V8l5 4-5 4z"
                    />
                  </g>
                </svg>
              </a>
            </div>
          </nav>
        </div>
      </div>
      <div
        class="col-md-12 row "
        id="bottom-section"
      >
        <div
          class="col-md-2 px-0"
          id="layout-sidemenu"
          style="height: 703px;"
        >
          <div
            class="list-group borderless mt-4 border-0 mx-0"
            id="sidebar"
          >
            <a
              aria-current="page"
              class="list-group-item list-group-item-action border-0 bg-transparent active"
              href="/"
              style="color: grey;"
            >
              <i
                class="fa fa-home"
              />
               Main
            </a>
            <a
              class="list-group-item list-group-item-action border-0 bg-transparent "
              href="/issues"
              style="color: rgb(140, 72, 159);"
            >
              <i
                class="fa fa-camera"
              />
               Issues
            </a>
            <a
              class="list-group-item list-group-item-action border-0 bg-transparent"
              href="/inspection"
              style="color: rgb(140, 72, 159);"
            >
              <i
                class="fa fa-music"
              />
               Inspection
            </a>
          </div>
        </div>
        <div
          class="col-md-10"
          id="layout-dashboard overflow-auto"
          style="height: 703px;"
        >
          <div>
            Nothing to see here
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
